### ISSUE TITLE
[Feature Research]: Lipsyncing Models

### ISSUE THREAD

**Issue by cobycloud** [swarmauri/swarmauri-sdk#1355]
### Feature Name

Lipsyncing Models

### Feature Description

Investigate methods for generating photo‑realistic lip movements in video, perfectly synchronized with arbitrary audio inputs.

### Research Findings

Evaluate key architectures such as Wav2Lip, LipGAN, and SyncNet for synchronization accuracy, video resolution support, real‑time performance, and ease of integration. Assess training/inference resource requirements and availability of pretrained models. Evaluate SOTA solutions. Evaluate pricing for cloud-based solution providers.

### Potential Impact

Supports features like video dubbing, avatar animation, and speech‑driven interactive characters, enhancing multimedia engagement in the SDK.

### Additional Resources (optional)

Prajwal et al., “Lip2Wav: High‑Quality Speech‑Driven Lip Animation” (2020)

Github: “Rudrabha/Wav2Lip”

### Feature Priority

Medium

**Comment by cobycloud** [comment]
Research triage update — RETAIN and narrow to a standalone provider/runtime package under swarmauri_video_lipsync_*. Before implementation, select one actively maintained runtime and record: license and model-weight terms, audio/video formats, sync-quality metric, identity preservation, CPU/GPU requirements, latency, offline behavior, and a deterministic fixture. Wav2Lip remains useful baseline evidence but is not by itself a current package contract: https://github.com/Rudrabha/Wav2Lip. Next step: amend this ticket with a selected runtime and acceptance matrix, then open one concrete standalone-package implementation.

**Comment by cobycloud** [comment]
## Research update — 2026-07-17

This remains a viable feature, but the implementation target should be narrowed. **Lip-sync editing** (regenerating the mouth region of an existing video) and **audio-driven talking-head generation** (animating a still image and changing pose/expression) are different capabilities and should not share one provider contract.

### Candidate matrix

| Model/runtime | Role | Current assessment | Package verdict |
|---|---|---|---|
| Wav2Lip | Existing-video lip-sync baseline | Reliable synchronization, but dated visual quality. The official pretrained implementation is restricted to personal/research/non-commercial use. | Keep only as research and regression evidence; do not ship its checkpoints. |
| LipGAN | Historical predecessor to Wav2Lip | Superseded and not a current production target. | Exclude from implementation. |
| SyncNet / StableSyncNet | Audio/video synchronization evaluator and training supervision | Not a generator. StableSyncNet reports 94% accuracy on HDTF and is useful for measuring output alignment. | Use as an evaluation metric/fixture tool, not as a provider. |
| VideoReTalking | Multi-stage existing-video editing | Apache-2.0 and technically viable, but older and operationally complex. | Reference implementation only for the first phase. |
| MuseTalk 1.5 | Local real-time lip-sync | 256x256 face region; reports 30+ FPS on Tesla V100. MIT code and model use permitted commercially, but requires a heavy Python 3.10/PyTorch/CUDA/MMLab/FFmpeg stack and documents identity-detail/jitter limitations. | Best local real-time candidate, implemented later through an isolated worker/runtime. |
| LatentSync 1.6 | Local high-quality batch lip-sync | Apache-2.0, 512x512, and reports better quality/sync/temporal results than Wav2Lip, VideoReTalking, and MuseTalk. Requires about 18 GB inference VRAM and 20-50 diffusion steps. | Best local quality candidate, implemented later through an isolated worker/runtime. |
| Sync Labs | Managed production API | Async jobs, cost estimation, webhooks, batches, video/image plus audio/text, and production model choices up to 4K. Current documented usage is about $0.04-$0.133 per output second, plus plan charges. | **Recommended first standalone provider.** |
| fal LatentSync | Managed open-model API | Simple queued video+audio endpoint without local GPU management. Current pricing is $0.20 for up to 40 seconds, then $0.005/output second. | Strong second provider; reuse existing Fal authentication/queue conventions. |

### Proposed package boundary

- `swarmauri_core.video_lipsync`: `ILipSync`, request/result types, capability metadata.
- `swarmauri_base.video_lipsync`: `LipSyncBase`, shared media validation and batch orchestration.
- `swarmauri_video_lipsync_synclabs`: first production provider.
- `swarmauri_video_lipsync_falai`: second API provider, initially supporting LatentSync.
- Later: `swarmauri_video_lipsync_musetalk` and `swarmauri_video_lipsync_latentsync` as worker/subprocess-backed local runtimes rather than installing their conflicting GPU dependency stacks into the SDK process.

The common operation should accept video and audio media references plus output path, model, duration/sync mode, and optional seed. The result should include the output URI/path, provider job ID, duration, FPS, model/version, processing time, provenance, and a synchronization score when available.

### Acceptance criteria for the first implementation

- Implement one API-backed standalone provider; Sync Labs is the recommended first target.
- Use a short consented or synthetic deterministic audio/video fixture.
- Validate playable output, preserved audio, duration behavior, and synchronization confidence/equivalent evidence.
- Test invalid media, mismatched durations, no detected face, provider failure, timeout, and async polling.
- Keep API keys and signed output URLs out of serialized component state.
- Record consent/provenance metadata and document identity/audio rights requirements.
- Do not bundle model weights in the SDK distribution.

### Primary evidence

- Wav2Lip repository and license restriction: https://github.com/Rudrabha/Wav2Lip
- MuseTalk repository, runtime, limitations, and license: https://github.com/TMElyralab/MuseTalk
- MuseTalk paper: https://arxiv.org/abs/2410.10122
- LatentSync repository, VRAM, and license: https://github.com/bytedance/LatentSync
- LatentSync comparisons and evaluation metrics: https://arxiv.org/html/2412.09262
- VideoReTalking: https://github.com/OpenTalker/video-retalking
- Sync Labs API: https://sync.so/docs/api-reference/api-overview
- Sync Labs billing: https://sync.so/docs/product/billing
- fal LatentSync endpoint: https://fal.ai/models/fal-ai/latentsync

**Research verdict:** retain this issue and replace the original Wav2Lip/LipGAN-centered implementation framing with a provider-neutral contract. The next concrete ticket should be a standalone `swarmauri_video_lipsync_synclabs` implementation linked back to this research issue.

**Comment by cobycloud** [comment]
## Example `swarmauri_video_*` package layout

A lip-sync implementation should follow the existing core → base → standalone-provider pattern used by the SDK. For the recommended first provider, the repository layout would be:

```text
pkgs/
├── core/
│   └── swarmauri_core/
│       └── video_lipsync/
│           ├── __init__.py
│           ├── ILipSync.py
│           ├── LipSyncRequest.py
│           └── LipSyncResult.py
├── base/
│   └── swarmauri_base/
│       └── video_lipsync/
│           ├── __init__.py
│           └── LipSyncBase.py
└── standards/
    └── swarmauri_video_lipsync_synclabs/
        ├── LICENSE
        ├── README.md
        ├── pyproject.toml
        ├── swarmauri_video_lipsync_synclabs/
        │   ├── __init__.py
        │   ├── SyncLabsLipSync.py
        │   ├── client.py
        │   └── retry.py
        └── tests/
            ├── unit/
            │   ├── test_sync_labs_lip_sync.py
            │   ├── test_payload_mapping.py
            │   └── test_serialization.py
            └── i9n/
                └── test_sync_labs_lip_sync_i9n.py
```

### Responsibilities

- `ILipSync.py`: provider-neutral `predict`, `apredict`, `batch`, and `abatch` contract.
- `LipSyncRequest.py`: video/audio references, model, output path, sync mode, seed, and provider options.
- `LipSyncResult.py`: output reference, provider job ID, model/version, duration, FPS, processing time, synchronization evidence, and provenance.
- `LipSyncBase.py`: shared media validation, batch concurrency, duration-policy validation, and safe serialization behavior.
- `SyncLabsLipSync.py`: public Swarmauri component; owns provider model selection and translates the common request/result types.
- `client.py`: private HTTP job submission, polling, cancellation, webhook/status mapping, and cost-estimation calls.
- `retry.py`: bounded retry/backoff for rate limits and transient provider failures.
- `tests/unit`: fully mocked and deterministic; no network, API key, real-person media, or billable generation.
- `tests/i9n`: opt-in live provider test using a short consented/synthetic fixture.

The provider package should expose one public component:

```toml
[project.entry-points.'swarmauri.video_lipsync']
SyncLabsLipSync = swarmauri_video_lipsync_synclabs:SyncLabsLipSync
```

Its runtime dependencies should remain small:

```toml
[project]
name = swarmauri_video_lipsync_synclabs
requires-python = >=3.10,<3.13
dependencies = [
    httpx>=0.27,
    pydantic>=2.0,
    swarmauri_core,
    swarmauri_base,
]
```

`pyproject.toml` and `README.md` must also carry the repository-standard planning classifier, Python 3.10–3.12 classifiers, Jacob Stewart author metadata, Swarmauri branding/badges, Features section, `uv`/`pip` installation, usage workflow, provider limitations, pricing caveat, consent/provenance guidance, and supported media constraints.

### Parallel providers

Additional providers should repeat only the standalone leaf package:

```text
pkgs/standards/
├── swarmauri_video_lipsync_synclabs/
├── swarmauri_video_lipsync_falai/
├── swarmauri_video_lipsync_musetalk/
└── swarmauri_video_lipsync_latentsync/
```

All four inherit the same `ILipSync`/`LipSyncBase` contract. `musetalk` and `latentsync` should adapt an isolated worker, container, or subprocess runtime; their CUDA/PyTorch/checkpoint stacks must not become dependencies of `swarmauri_core`, `swarmauri_base`, or the umbrella `swarmauri` package.

### Repository integration

The implementation PR must also update workspace membership/lock data and component discovery mappings where required. Provider secrets, downloaded weights, temporary frames, generated media, and signed output URLs must remain outside package source and serialized component state.

**Comment by cobycloud** [comment]
## Revised contract and `SyncLabsLipSync` proposal

This supersedes the earlier package-tree proposal that introduced `LipSyncRequest.py` and `LipSyncResult.py`.

### Contract correction

Separate request/result objects are not currently the convention for Swarmauri media families:

- `ITextToSpeech` accepts ordinary arguments and returns output-path strings: https://github.com/swarmauri/swarmauri-sdk/blob/ea57b506259ab9081cde2ba178b9071ddd66232f/pkgs/core/swarmauri_core/tts/ITextToSpeech.py
- `IGenImage` exposes direct generation methods without request/result wrappers: https://github.com/swarmauri/swarmauri-sdk/blob/ea57b506259ab9081cde2ba178b9071ddd66232f/pkgs/core/swarmauri_core/image_gens/IGenImage.py
- The standalone PlayHT provider demonstrates the current Pydantic component, `SecretStr`, sync/async, and output-path pattern: https://github.com/swarmauri/swarmauri-sdk/blob/ea57b506259ab9081cde2ba178b9071ddd66232f/pkgs/standards/swarmauri_tts_playht/swarmauri_tts_playht/PlayHTModel.py

The first lip-sync contract should therefore use ordinary arguments and return `str`. Provider job payloads remain private until multiple providers establish a genuinely shared metadata contract.

```python
class ILipSync(ABC):
    @abstractmethod
    def predict(
        self,
        video_url: str,
        audio_url: str,
        output_path: str = "output.mp4",
        **kwargs: Any,
    ) -> str: ...

    @abstractmethod
    async def apredict(
        self,
        video_url: str,
        audio_url: str,
        output_path: str = "output.mp4",
        **kwargs: Any,
    ) -> str: ...
```

### Revised layout

```text
pkgs/core/swarmauri_core/video_lipsync/
??? ILipSync.py
??? __init__.py

pkgs/base/swarmauri_base/video_lipsync/
??? LipSyncBase.py
??? __init__.py

pkgs/standards/swarmauri_video_lipsync_synclabs/
??? LICENSE
??? README.md
??? pyproject.toml
??? swarmauri_video_lipsync_synclabs/
?   ??? SyncLabsLipSync.py
?   ??? retry.py
?   ??? __init__.py
??? tests/
    ??? unit/
    ?   ??? test_sync_labs_lip_sync.py
    ?   ??? test_payload_mapping.py
    ?   ??? test_secret_serialization.py
    ??? i9n/
        ??? test_sync_labs_lip_sync_i9n.py
```

### Representative `SyncLabsLipSync.py`

This first adapter accepts remote URLs. Local files can later use Sync's asset-upload endpoints without changing the public return type.

```python
import time
from pathlib import Path
from typing import Any, Literal

import httpx
from pydantic import Field, PrivateAttr, SecretStr

from swarmauri_base.ComponentBase import ComponentBase
from swarmauri_base.video_lipsync.LipSyncBase import LipSyncBase


@ComponentBase.register_type(LipSyncBase, "SyncLabsLipSync")
class SyncLabsLipSync(LipSyncBase):
    allowed_models: list[str] = [
        "lipsync-2",
        "lipsync-2-pro",
        "lipsync-2-mini",
        "sync-3",
    ]
    api_key: SecretStr = Field(exclude=True, repr=False)
    name: str = "lipsync-2"
    type: Literal["SyncLabsLipSync"] = "SyncLabsLipSync"
    base_url: str = "https://api.sync.so"
    request_timeout: float = 60.0
    generation_timeout: float = 900.0
    poll_interval: float = 1.0

    _failed_statuses: frozenset[str] = PrivateAttr(
        default=frozenset({"FAILED", "REJECTED"})
    )

    def _headers(self) -> dict[str, str]:
        return {
            "x-api-key": self.api_key.get_secret_value(),
            "accept": "application/json",
            "content-type": "application/json",
        }

    def _wait_for_output(
        self, client: httpx.Client, generation_id: str
    ) -> str:
        deadline = time.monotonic() + self.generation_timeout
        while time.monotonic() < deadline:
            response = client.get(
                f"/v2/generate/{generation_id}",
                params={"wait": 10},
            )
            response.raise_for_status()
            generation = response.json()
            status = generation["status"]

            if status == "COMPLETED":
                output_url = generation.get("outputUrl")
                if not output_url:
                    raise RuntimeError("Completed job has no outputUrl")
                return output_url
            if status in self._failed_statuses:
                code = generation.get("errorCode", "unknown_error")
                message = generation.get("error", "No provider message")
                raise RuntimeError(f"Sync generation failed: {code}: {message}")
            time.sleep(self.poll_interval)

        raise TimeoutError(f"Generation {generation_id} timed out")

    @staticmethod
    def _download(output_url: str, output_path: str) -> str:
        destination = Path(output_path).expanduser().resolve()
        destination.parent.mkdir(parents=True, exist_ok=True)

        # A separate unauthenticated client prevents disclosure of the Sync
        # API key when outputUrl points to a different host.
        with httpx.stream(
            "GET", output_url, timeout=120.0, follow_redirects=True
        ) as response:
            response.raise_for_status()
            with destination.open("wb") as output_file:
                for chunk in response.iter_bytes():
                    output_file.write(chunk)
        return str(destination)

    def predict(
        self,
        video_url: str,
        audio_url: str,
        output_path: str = "output.mp4",
        *,
        sync_mode: str = "cut_off",
        **kwargs: Any,
    ) -> str:
        payload = {
            "model": self.name,
            "input": [
                {"type": "video", "url": video_url},
                {"type": "audio", "url": audio_url},
            ],
            "options": {"sync_mode": sync_mode, **kwargs},
            "outputFileName": Path(output_path).name,
        }

        with httpx.Client(
            base_url=self.base_url,
            headers=self._headers(),
            timeout=self.request_timeout,
        ) as client:
            response = client.post("/v2/generate", json=payload)
            response.raise_for_status()
            generation_id = response.json()["id"]
            output_url = self._wait_for_output(client, generation_id)

        return self._download(output_url, output_path)

    async def apredict(
        self,
        video_url: str,
        audio_url: str,
        output_path: str = "output.mp4",
        *,
        sync_mode: str = "cut_off",
        **kwargs: Any,
    ) -> str:
        # Implement with AsyncClient, asyncio.sleep, and aiter_bytes;
        # do not call the blocking predict method from async code.
        ...
```

### Required production completion

- Implement native-async `apredict`; never delegate to blocking `predict`.
- Add local-file support through `POST /v2/assets/upload`, byte upload, and asset registration.
- Retry `429` and transient `5xx` responses with bounded backoff and `Retry-After` support.
- Validate model names and options before submitting billable work.
- Download `outputUrl` with a separate unauthenticated client to prevent cross-host API-key disclosure.
- Prove through tests that `model_dump()` excludes `api_key`.
- Keep live tests opt-in and use a short consented/synthetic fixture.

### Provider evidence

The current Sync Labs contract documents `POST /v2/generate`, `GET /v2/generate/{id}`, `x-api-key` authentication, one visual input plus audio/text, terminal states `COMPLETED`/`FAILED`/`REJECTED`, `outputUrl`, and long polling with `wait=1..10`.

- API overview: https://sync.so/docs/api-reference/api-overview
- OpenAPI 3.1 specification: https://sync.so/openapi.json
- Billing and cost estimation: https://sync.so/docs/product/billing

**Updated verdict:** implement the minimal core/base contract and `swarmauri_video_lipsync_synclabs`; do not add `LipSyncRequest` or `LipSyncResult` in the first implementation.

**Comment by cobycloud** [comment]
Implementation is now available in draft PR #6216: https://github.com/swarmauri/swarmauri-sdk/pull/6216

This first slice adds the shared `ILipSync` and queued-job contracts, `LipSyncBase`, normalized job events, and the standalone `swarmauri_video_lipsync_synclabs` provider. The provider supports public URLs and local asset uploads, synchronous/asynchronous prediction, progress observation, cost estimation, retry handling, credential-isolated uploads/downloads, and streamed final-artifact downloads.

`stream`/`astream` are intentionally not implemented for Sync Labs because its API exposes queued progress plus a completed artifact, not incremental output-video bytes. A genuine streaming capability remains deferred to a runtime that can prove incremental media generation.

Evidence:
- PR and checks: https://github.com/swarmauri/swarmauri-sdk/pull/6216
- Sync generation contract: https://sync.so/docs/api-reference/api/generate-api/create
- Sync status/progress contract: https://sync.so/docs/api-reference/api/generate-api/get
- Sync local asset upload flow: https://sync.so/docs/developer-guides/asset-uploads

**Comment by cobycloud** [comment]
Closing as completed by [PR #6216](https://github.com/swarmauri/swarmauri-sdk/pull/6216), merged as commit [26d0310e](https://github.com/swarmauri/swarmauri-sdk/commit/26d0310ea3f2dbe4c8de356faccdb648dc448dc2).

Delivered evidence:
- provider-neutral `ILipSync` and queued-job contracts
- `LipSyncBase` and the `VideoLipSync` resource type
- standalone `swarmauri_video_lipsync_synclabs`
- sync/async generation, local asset uploads, normalized progress events, cost estimation, retry handling, secret-safe serialization, and credential-isolated artifact downloads
- passing core, base, provider, umbrella, metadata, and license workflows

Additional Fal or local GPU runtimes should be tracked as separate concrete provider tickets rather than keeping this research ticket open.
